lib/scenario.js

```diff
@@ -80,45 +80,83 @@ module.exports.test = (test) => {
  * Injects arguments to function from controller
  */
 module.exports.injected = function (fn, suite, hookName) {
-  return function () {
+  return function (done) {
+    recorder.errHandler(function (err) {
+      recorder.session.start('teardown');
+      event.emit(event.test.failed, suite, err);
+      if (hookName === 'after') event.emit(event.test.after, suite);
+      if (hookName === 'afterSuite') event.emit(event.suite.after, suite);
+      recorder.add(() => done(err));
+    });
     try {
       recorder.startUnlessRunning();
       this.test.body = fn.toString();
       fn.apply(this, getInjectedArguments(fn));
     } catch (err) {
       recorder.throw(err);
     }
-    recorder.catch((err) => {
-      event.emit(event.test.failed, suite, err); // emit
-      if (hookName === 'after') event.emit(event.test.after, suite);
-      if (hookName === 'afterSuite') event.emit(event.suite.after, suite);
-      throw err;
-    });
-    return recorder.promise();
+    recorder.add(`finish ${hookName} hook`, () => done());
+    recorder.catch();
   };
 };
 
 /**
  * Starts promise chain, so helpers could enqueue their hooks
  */
-module.exports.setup = function () {
-  recorder.startUnlessRunning();
-  event.emit(event.test.before);
+module.exports.setup = function (suite) {
+  try {
+    recorder.startUnlessRunning();
+    event.emit(event.test.before);
+  } catch (err) {
+    recorder.throw(err);
+  }
+  recorder.catch((err) => {
+    event.emit(event.test.failed, suite, err);
+    throw err;
+  });
+  return recorder.promise();
 };
 
-module.exports.teardown = function () {
-  recorder.startUnlessRunning();
-  event.emit(event.test.after);
+module.exports.teardown = function (suite) {
+  try {
+    recorder.startUnlessRunning();
+    event.emit(event.test.after);
+  } catch (err) {
+    recorder.throw(err);
+  }
+  recorder.catch((err) => {
+    event.emit(event.test.failed, suite, err);
+    throw err;
+  });
+  return recorder.promise();
 };
 
 module.exports.suiteSetup = function (suite) {
-  recorder.startUnlessRunning();
-  event.emit(event.suite.before, suite);
+  try {
+    recorder.startUnlessRunning();
+    event.emit(event.suite.before, suite);
+  } catch (err) {
+    recorder.throw(err);
+  }
+  recorder.catch((err) => {
+    event.emit(event.test.failed, suite, err);
+    throw err;
+  });
+  return recorder.promise();
 };
 
 module.exports.suiteTeardown = function (suite) {
-  recorder.startUnlessRunning();
-  event.emit(event.suite.after, suite);
+  try {
+    recorder.startUnlessRunning();
+    event.emit(event.suite.after, suite);
+  } catch (err) {
+    recorder.throw(err);
+  }
+  recorder.catch((err) => {
+    event.emit(event.test.failed, suite, err);
+    throw err;
+  });
+  return recorder.promise();
```

Comment:
still the same: please make one function and reuse it for suiteTeardown, suiteSetup and setup